### PR TITLE
 Update rules_go, repo-infra, and rules_docker dependencies

### PR DESCRIPTION
**Pull request by ixdy**
**What this PR does / why we need it**: several of our bazel dependencies were getting pretty old, since they required bazel 0.5.4+ but there were various failures if we tried to use them with bazel 0.5.4.

Now that bazel 0.6.0 (and 0.6.1) have been out for a while, we can bump our dependencies and get a number of fixes and new functionality.

x-ref #52677 and others

**Special notes for your reviewer**:
* This will now explicitly require bazel 0.6.0+ to build kubernetes.
* Our staging directories are causing some issues for `gazelle`; it wants to set `importpath = "k8s.io/kubernetes/staging/src/k8s.io/blah"` instead of `importpath = "k8s.io/blah"`. I'm not sure what is the correct way to fix this; what we're doing here is pretty weird and nonstandard. I've used a `sed` substitution for now.
* The `-proto=default` option of `gazelle` has a number of bugs right now (https://github.com/bazelbuild/rules_go/issues/888, https://github.com/bazelbuild/rules_go/issues/900, https://github.com/bazelbuild/rules_go/issues/907), so I am forcing the legacy behavior.

**Release note**:

```release-note
NONE
```

/assign @mikedanese @spxtr @BenTheElder

### CODE REVIEW

**Comment by ixdy** [comment]
I supposed I could add functionality to `kazel` to fixup the `staging/` paths rather than using `sed`. I'm not sure which is worse.

**Comment by BenTheElder** [comment]
This looks fine to me, though I'm also not sure which is worse. 
/lgtm

**Comment by ixdy** [comment]
/assign @thockin 
for "update-every-`BUILD`-file" approval

I left this PR as three separate commits to make it easier to understand what's changing:
* Update all the deps and impose a bazel 0.6.0+ check
* Fix `hack/{update,verify}-bazel.sh` to work with newer `gazelle`
* Run `hack/update-bazel.sh`

The new `importpath` field is due to some cleanup work in `rules_go` (https://github.com/bazelbuild/rules_go/issues/721). It's a bit of a pain right now (since we have to update basically *every* `BUILD` file), but will clean up some of the rules later (e.g. using a better name than "go_default_library"). The good news is that `gazelle` did most of the hard work for us.

I'm still not happy about the `staging/` `sed` hack, but `staging/` seems like a bit of a hack already, so...

File: hack/update-bazel.sh [hunk 1]
```diff
@@ -37,5 +37,13 @@ touch "${KUBE_ROOT}/vendor/BUILD"
...
     -build_file_name=BUILD,BUILD.bazel \
     -external=vendored \
+    -proto=legacy \
     -mode=fix
+# gazelle gets confused by our staging/ directory, prepending an extra
+# "k8s.io/kubernetes/staging/src" to the import path.
+# gazelle won't follow the symlinks in vendor/, so we can't just exclude
+# staging/. Instead we just fix the bad paths with sed.
+find staging -name BUILD -o -name BUILD.bazel | \
+  xargs sed -i 's|\(importpath = "\)k8s.io/kubernetes/staging/src/\(.*\)|\1\2|'
```
Comment (cblecker) [inline hack/update-bazel.sh]: Yeah, the sed here isn't pretty, but we have to do something similar already to update staging godeps:
https://github.com/kubernetes/kubernetes/blob/97e002352f459aebf7344babf707837b4f62be2f/hack/update-staging-godeps-dockerized.sh#L79

File: hack/verify-bazel.sh [hunk 2]
```diff
@@ -31,20 +31,24 @@ fi
...
-kazel_diff=$(kazel -dry-run -print-diff -root="$(kube::realpath ${KUBE_ROOT})")
-
-if [[ -n "${gazelle_diff}" || -n "${kazel_diff}" ]]; then
-  echo "${gazelle_diff}"
-  echo "${kazel_diff}"
+_tmpdir="$(mktemp -d -t verify-bazel.XXXXXX)"
+function cleanup {
+  rm -rf "${_tmpdir}"
+}
+trap cleanup EXIT
```
Comment (cblecker) [inline hack/verify-bazel.sh]: nit: could use `kube::util::trap_add`, but that's only required if we cascade scripts and need to call multiple traps.

**Review by cblecker — APPROVED** [review]
Yeah, I'm not super jazzed about the sed, but it also isn't too horrible either. Especially when compared with some of the other staging hacks.

File: hack/verify-bazel.sh (hunk 2 above)
Comment (ixdy) [reply hack/verify-bazel.sh]: good suggestion, done.

**Comment by BenTheElder** [comment]
@ixdy picked up conflicts with 
```
cmd/kubeadm/app/node/BUILD
pkg/controller/certificates/BUILD
```
:/

**Comment by ixdy** [comment]
rebased and regenerated BUILD files. going to self-LGTM so this hopefully merges before it breaks again.

**Pull request merged**